_posts/2018-09-06-fullstack-typed-api-with-typescript.md

```diff
@@ -0,0 +1,427 @@
+---
+layout: post
+title: Fullstack shared typed API with TypeScript
+author: jgke
+excerpt: >
+  TypeScript makes it possible to share type information between frontend and
+  backend. Here's how to use it to ensure consistent API calls between the
+  frontend and backend.
+tags:
+ - TypeScript
+ - Full stack
+---
+
+Did you get yet another error due to the JavaScript frontend slipping an
+unnoticed ``undefined`` to a request or passing a string instead of a number,
+which causes requests to fail without a clear reason? Or even worse, succeed in
+a weird way? Maybe you changed the API on the server side but forgot to update
+the client? Consider sharing the API as a single type between the
+frontend and the backend. This blog post explains the process on how to build a
+shared API using fullstack TypeScript. As a bonus, you also get runtime type
+checks!
+
+## Sketching out an API
+
+Let's make a simple API:
+- ``GET /customers``: fetch a list of customers
+- ``POST /customers``: create a new customer
+- ``GET /search?id=123&name=foobar``: search a customer by id/name
+
+As we're building a JSON-based API, every request and return value has to be an
+object. We can handle this transparently by defining the format for responses:
+
+```ts
+export interface ApiResponse<Res> {
+    message: Res;
+}
+```
+
+All HTTP requests look roughly like functions ``(queryParameters, body) =>
+returnvalue`` (for GETs, this means (queryParameters, {}) => returnValue``).
+This means that we can model the API with the following type:
+
+```ts
+type Customer = {
+    id: number,
+    name: string
+}
+type ApiMap = {
+    customers: {
+        GET: () => Promise<Customer[]>,
+        POST: (customerName: string) => Promise<Customer>
+    },
+    search: {
+        GET: (query: {id?: number, name?: string}) => Promise<Customer[]>
+    }
+}
+```
+
+On the client side, the ApiMap type can be used to define a object:
+
+
+```ts
+function fetchAny(url: string, queryParams?: any,
+                  method?: string, body?: any): Promise<any> {
+    let address = new URL(url);
+    Object.keys(queryParams || {})
+        .forEach(key => address.searchParams.append(key, queryParams[key]));
+    return fetch(address.toString(), {method, body})
+        .then(response => response.json()
+            .then((responseBody: any) => new Promise<Res>(resolve =>
+                resolve(responseBody.message))));
+}
+
+const api: ApiMap = {
+    customers: {
+        GET: () => fetchAny('/customers'),
+        POST: (customerName: string) => fetchAny('/customers', {}, 'POST', customerName)
+    },
+    search: {
+        GET: (params: {id: number?, name: string?}) => fetchAny('/search', params, 'PUT')
+    }
+};
+
+// ...
+
+api.customers.GET().then(val => /* ... */ val)
+```
+
+The client side implementation can now be used as if the API was just an object
```

Comment:
client side -> client-side

"You can now use…" instead of "can now be used" is a bit less stilted and makes the agency more clear, but this isn't wrong.